tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -139,8 +153,335 @@ impl ServerKey {
+                        let previous_block = carry_out_slice.get(previous_index) as *const _;
+
+                        self.key.unchecked_apply_lookup_table_bivariate_assign(
+                            &mut *current_block,
+                            &*previous_block,
+                            &lut_carry_propagation_sum,
+                        );
+                    }
+                });
+        }
+
+        // Down-Sweep phase
+        let mut buffer = Vec::with_capacity(num_blocks / 2);
+        self.key
+            .create_trivial_assign(&mut carry_out[num_blocks - 1], 0);
+        for i in (0..num_steps).rev() {
+            let i_plus_1_pow_2 = 2usize.checked_pow((i + 1) as u32).unwrap();
+            let i_pow_2 = 2usize.checked_pow(i as u32).unwrap();
+
+            (0..num_blocks)
+                .into_par_iter()
+                .step_by(i_plus_1_pow_2)
+                .map(|k| {
+                    // Since our carry_propagation LUT ie sum function
+                    // is not commutative we have to reverse operands
+                    self.key.unchecked_apply_lookup_table_bivariate(
+                        &carry_out[k + i_pow_2 - 1],
+                        &carry_out[k + i_plus_1_pow_2 - 1],
+                        &lut_carry_propagation_sum,
+                    )
+                })
+                .collect_into_vec(&mut buffer);
+
+            let mut drainer = buffer.drain(..);
+            for k in (0..num_blocks).step_by(i_plus_1_pow_2) {
+                let b = drainer.next().unwrap();
+                carry_out.swap(k + i_pow_2 - 1, k + i_plus_1_pow_2 - 1);
+                carry_out[k + i_plus_1_pow_2 - 1] = b;
+            }
+            drop(drainer);
+            assert!(buffer.is_empty());
+        }
+
+        // The first step of the Down-Sweep phase sets the
+        // first block to 0
+        lhs.blocks
+            .par_iter_mut()
+            .zip(carry_out.par_iter())
+            .for_each(|(block, carry_in)| {
+                self.key.unchecked_add_assign(block, carry_in);
+                self.key.message_extract_assign(block);
+            });
+    }
+
+    /// Initialization function for parallal carryless sum
+    ///
+    /// This function adds rhs into lhs
+    ///
+    /// It Returns the array that tells whether the
+    /// block resulting from the sum will propagat or generate
+    /// a carry. A Prefix sum / cumulative sum then needs to be done
+    /// to get the final carry that the block will output
+    fn add_and_generate_init_carry_array<PBSOrder: PBSOrderMarker>(
+        &self,
+        lhs: &mut RadixCiphertext<PBSOrder>,
+        rhs: &RadixCiphertext<PBSOrder>,
+        add_extra_one: AddExtraOne,
+    ) -> Vec<crate::shortint::CiphertextBase<PBSOrder>> {
+        let modulus = self.key.message_modulus.0 as u64;
+
+        let lut_does_block_generate_carry =
+            self.key.generate_accumulator(|x| (x >= modulus) as u64);
```

Comment:
kind of dangerous here to not do some matching to return the same values as the OutputCarry enum, here they match but it would be easy to make a mistake it feels like

Reply:
ok now I see how the Propagate variant is a smart trick that is not a true carry value, still I feel like there should be consistency between both accumulators potentially :/

Reply:
yeah I use enum variant to have more readability

I updated this part to use the variant